contracts/statechannels/GRTWithdrawHelper.sol

```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.7.3;
+pragma experimental ABIEncoderV2;
+
+import "../staking/IStaking.sol";
+import "../token/IGraphToken.sol";
+
+import "./WithdrawHelper.sol";
+
+/**
+ * @title WithdrawHelper contract for GRT tokens
+ * @notice This contract encodes the logic that connects the withdrawal of funds from a channel
+ * back to the protocol for a particular allocation.
+ */
+contract GRTWithdrawHelper is WithdrawHelper {
+    struct CollectData {
+        address staking;
+        address allocationID;
+    }
+
+    // -- State --
+
+    address public tokenAddress;
+
+    /**
+     * @notice Contract constructor.
+     * @param _tokenAddress Token address to use for transfers
+     */
+    constructor(address _tokenAddress) {
+        tokenAddress = _tokenAddress;
+    }
+
+    /**
+     * @notice Returns the ABI encoded representation of a CollectData struct.
+     * @param _collectData CollectData struct with information about how to collect funds
+     */
+    function getCallData(CollectData calldata _collectData) public pure returns (bytes memory) {
+        return abi.encode(_collectData);
+    }
+
+    /**
+     * @notice Execute hook used by a channel to send funds to the protocol.
+     * @param _wd WithdrawData struct for the withdrawal commitment
+     * @param _actualAmount Amount to transfer to the Staking contract
+     */
+    function execute(WithdrawData calldata _wd, uint256 _actualAmount) external override {
+        require(_wd.assetId == tokenAddress, "GRTWithdrawHelper: !token");
+
+        // Decode and validate collect data
+        CollectData memory collectData = abi.decode(_wd.callData, (CollectData));
+        require(collectData.staking != address(0), "GRTWithdrawHelper: !staking");
+        require(collectData.allocationID != address(0), "GRTWithdrawHelper: !allocationID");
+
+        // Approve the staking contract to pull the transfer amount
+        require(
+            IGraphToken(_wd.assetId).approve(collectData.staking, _actualAmount),
+            "GRTWithdrawHelper: !approve"
+        );
+
+        // Call the staking contract to collect funds from this contract
+        IStaking(collectData.staking).collect(_actualAmount, collectData.allocationID);
```

Comment:
If this function reverts, the channel funds will be gone. We need to try-catch this call and provide a backup address the funds can be sent to (can even be the channel multisig address which is already part of the withdraw data).

Reply:
The collect function will only revert if: the allocation ID passed id 0x0 or invalid, it will then accept funds in any case.

My question is, why is that those funds will be lost? won't the collect call reverting make the transaction fail?

I based this on how the https://github.com/connext/vector-withdraw-helpers/blob/main/contracts/UniswapWithdrawHelper/UniswapWithdrawHelper.sol works.

Reply:
The transaction will fail, but the funds are already signed out of the channel offchain, so they are effectively marked as "used".

Reply:
@rhlsthrm and if the transfer to the channel is made instead, it will be credited to Bob, even though it might've being made by Alice?

Reply:
@rhlsthrm what happens if this transaction fails due to an out of gas exception?

Reply:
> @rhlsthrm and if the transfer to the channel is made instead, it will be credited to Bob, even though it might've being made by Alice?

Yes, if the transfer is made directly to the channel address, it will be credited to Bob. However, you can call `depositAlice`. We might be able to program the logic to send to channel vs call `depositAlice` into the helper itself.

Reply:
> @rhlsthrm what happens if this transaction fails due to an out of gas exception?

If it fails with hitting the gas limit on the transaction, you can resubmit the withdrawal commitment with a higher gas limit. It's only the case of a _permanent_ failure that the funds are unrecoverable.

Reply:
Perfect